In [1]:
import pandas as pd
import numpy as np

In [2]:
ttino = pd.read_csv('tarantino.csv')
ttino.head()

movie  type      word  minutes_in
0  Reservoir Dogs  word      dick        0.40
1  Reservoir Dogs  word     dicks        0.43
2  Reservoir Dogs  word    fucked        0.55
3  Reservoir Dogs  word   fucking        0.61
4  Reservoir Dogs  word  bullshit        0.61

Проверим, есть ли пропущенные значения.

In [10]:
ttino.isnull().sum()

movie           0
type            0
word          190
minutes_in      0
dtype: int64

Найдено 190 пропущенных значений в столбце word.

Посмотрим, какие значения присутствуют в столбце type, чтобы определить, каким type соответствуют пропущенные значения из столбца word.

In [15]:
ttino['type'].value_counts()

type
word     1704
death     190
Name: count, dtype: int64

В столбце type присутствуют только два уникальных значения: word в 1704 строчках и death в 190. 

Так как количество строк с type death равно количеству строк с пропущенным значением в столбце word, то можно сделать предположение, что это одни и те же строки, то есть чт именно в строчках с type death не прописано значение в столбце word. Проверим это:

In [29]:
ttino[ttino['type'] == 'death'].isnull().sum()

movie           0
type            0
word          190
minutes_in      0
dtype: int64

Получили, что в 190 строчках с type death присутсвует пропущенное значение в столбце word. Таким образом, это значит, что пропущенные значения есть только в type death, а в свою очередь, в type word такие значения отсутсвуют. Следовательно, мы знаем каждое конкретное слово.

Итак, перейдем к первому пункту задания и определим число летальных исходов, проклятий и их соотношение в каждом отдельном фильме:

In [31]:
agg_df = ttino.groupby(['movie', 'type']).size().unstack(fill_value=0)
agg_df['total'] = agg_df.sum(axis=1)

agg_df['death_percentage'] = round(100.0 * agg_df['death'] / agg_df['total'],1)
agg_df['word_percentage'] = round(100.0 * agg_df['word'] / agg_df['total'],1)

agg_df

type                 death  word  total  death_percentage  word_percentage
movie                                                                     
Django Unchained        47   262    309              15.2             84.8
Inglorious Basterds     48    58    106              45.3             54.7
Jackie Brown             4   368    372               1.1             98.9
Kill Bill: Vol. 1       63    57    120              52.5             47.5
Kill Bill: Vol. 2       11    69     80              13.8             86.2
Pulp Fiction             7   469    476               1.5             98.5
Reservoir Dogs          10   421    431               2.3             97.7

Далее, определим частоту употребления конкретных проклятий, то есть выведем, сколько раз сказано каждое проклятие и сколько это составляет в процентах от общего числа проклятий:

In [8]:
agg_df_words = ttino[ttino['type'] == 'word'].groupby('word', as_index = False).agg(amount = ('word', 'count'))


agg_df_words['total'] = agg_df_words['amount'].sum()
agg_df_words['word_percentage'] = round(100.0 *agg_df_words['amount'] / agg_df_words['total'],1)
agg_df_words = agg_df_words[['word', 'amount', 'word_percentage']]

agg_df_words

word  amount  word_percentage
0             ass     140              8.2
1           asses       4              0.2
2         asshead       1              0.1
3         asshole      10              0.6
4        assholes       5              0.3
5         bastard      10              0.6
6        bastards       1              0.1
7           bitch      64              3.8
8         bitches       9              0.5
9        bullshit      15              0.9
10    chickenshit       1              0.1
11   cockblockery       1              0.1
12     cocksucker       2              0.1
13           cunt       2              0.1
14          cunts       1              0.1
15           damn      37              2.2
16         damned       3              0.2
17           dick      14              0.8
18       dickless       1              0.1
19          dicks       5              0.3
20        dumbass       1              0.1
21         faggot       1              0.1
22           fuck     213             12.5
23         fucked      25              1.5
24         fucker       8              0.5
25        fuckers       1              0.1
26       fuckface       1              0.1
27       fuckhead       1              0.1
28        fucking     407             23.9
29          fucks       3              0.2
30         fuckup       1              0.1
31        goddamn     113              6.6
32      goddamned       1              0.1
33           gook       1              0.1
34          gooks       2              0.1
35           hell      45              2.6
36       horeshit       1              0.1
37      horseshit       2              0.1
38        jackass       1              0.1
39            jap       1              0.1
40           japs       2              0.1
41     jew (verb)       1              0.1
42          merde       1              0.1
43   motherfucker      70              4.1
44  motherfuckers      15              0.9
45  motherfucking      27              1.6
46        n-word      179             10.5
47         negro        5              0.3
48          pussy      11              0.6
49           shit     221             13.0
50       shithead       3              0.2
51       shitless       1              0.1
52       shitload       2              0.1
53      shittiest       1              0.1
54       shitting       1              0.1
55         shitty       4              0.2
56          slope       1              0.1
57           slut       1              0.1
58          squaw       1              0.1
59        wetback       2              0.1

Затем, определим распределение времени между проклятиями, то есть медиану времени, а также квантиль 0.25, 0.75:

In [39]:
agg_df_q = ttino[ttino['type'] == 'word'].groupby('word', as_index = False).agg(q_25_minutes = ('minutes_in', lambda x: x.quantile(0.25)),
           median_minutes = ('minutes_in', lambda x: x.quantile(0.5)),
           q_75_minutes = ('minutes_in', lambda x: x.quantile(0.75)))
agg_df_q = agg_df_q[['word', 'q_25_minutes', 'median_minutes', 'q_75_minutes']]

agg_df_q

word  q_25_minutes  median_minutes  q_75_minutes
0             ass       36.6100          76.820      116.3800
1           asses       33.0625          76.565      123.9475
2         asshead       42.8600          42.860       42.8600
3         asshole       26.5200          34.370       46.1475
4        assholes       35.4800          48.480       55.8300
5         bastard       47.2550          56.505       82.8775
6        bastards       44.1200          44.120       44.1200
7           bitch       31.9975          75.015      111.7725
8         bitches       64.3200          77.360      115.0700
9        bullshit       28.4150          54.250       85.2300
10    chickenshit       29.9200          29.920       29.9200
11   cockblockery        9.6200           9.620        9.6200
12     cocksucker       37.5250          55.320       73.1150
13           cunt       51.2175          74.005       96.7925
14          cunts       35.1500          35.150       35.1500
15           damn       45.4000          77.470      111.5300
16         damned       66.6150          67.500       81.4250
17           dick        1.7500           1.750        4.6525
18       dickless       29.6300          29.630       29.6300
19          dicks        1.5600           1.750       35.2200
20        dumbass      135.7200         135.720      135.7200
21         faggot       81.6500          81.650       81.6500
22           fuck       32.5500          63.330      110.0000
23         fucked       15.8800          44.650       75.8700
24         fucker       31.5825          43.625       73.8300
25        fuckers        2.2800           2.280        2.2800
26       fuckface       62.3500          62.350       62.3500
27       fuckhead       80.7000          80.700       80.7000
28        fucking       28.6600          57.850      102.1000
29          fucks        4.4200           6.660       18.6450
30         fuckup       40.4800          40.480       40.4800
31        goddamn       29.0500          76.200      117.8500
32      goddamned       59.6500          59.650       59.6500
33           gook        2.2800           2.280        2.2800
34          gooks       18.6475          34.875       51.1025
35           hell       32.1100          51.030       90.3000
36       horeshit       68.4000          68.400       68.4000
37      horseshit      141.3675         147.305      153.2425
38        jackass      120.6500         120.650      120.6500
39            jap        2.9300           2.930        2.9300
40           japs       23.9475          32.365       40.7825
41     jew (verb)       17.3300          17.330       17.3300
42          merde       82.7500          82.750       82.7500
43   motherfucker       21.1350          95.105      130.3450
44  motherfuckers       22.0550         113.050      136.0000
45  motherfucking       16.1900          53.170      124.7250
46        n-word        28.2150          79.800      110.6400
47         negro        58.4300          59.380       59.5700
48          pussy       32.6350          63.810       79.1050
49           shit       28.8300          69.780      116.7000
50       shithead       73.3250         121.530      121.5750
51       shitless      111.9200         111.920      111.9200
52       shitload       49.0625          49.525       49.9875
53      shittiest       30.7700          30.770       30.7700
54       shitting       34.9000          34.900       34.9000
55         shitty       27.1275          30.535       34.9650
56          slope       67.5000          67.500       67.5000
57           slut       90.5300          90.530       90.5300
58          squaw       73.0000          73.000       73.0000
59        wetback       22.4900          41.310       60.1300